File: 动态规划/1.爬楼梯/2.记忆化dfs/python/main.py
from typing import List


class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n
        map1 = {}
        
        def dfs(n):
            if n<=1:
                return 1
            if n in map1:
                return map1[n]
            ans = dfs(n-1)+dfs(n-2)
            map1[n] = ans 
            return ans 

        return dfs(n)


if __name__ == '__main__':
    s = Solution()

File: 动态规划/1.爬楼梯/2.记忆化dfs/java/Main.java
import java.util.*;
public class Main {

    Map<Integer,Integer> map1 = new HashMap<>();
    public int dfs(int index){
        if(index<=1){
            return 1;
        }
        if(map1.containsKey(index)){
            return map1.get(index);
        }
        int ans = dfs(index-1)+dfs(index-2);
        map1.put(index,ans);
        return ans;
    }
    public int climbStairs(int n) {
        if(n<=2){
            return n;
        }
        return dfs(n);
    }
    public static void main(String[] args) {
        Main main = new Main();
    }
}